### PR TITLE
py-pystan: Add new package

### PR DESCRIPTION
**Pull request by haralmha**

### CODE REVIEW

**Comment by haralmha** [comment]
Depends on py-marshmallow, py-webargs, py-pysimdjson and py-httpstan, which are not yet added to spack. PRs made here: https://github.com/spack/spack/pull/29764, https://github.com/spack/spack/pull/29768, https://github.com/spack/spack/pull/29767 and https://github.com/spack/spack/pull/29770.

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    """PyStan is a Python interface to Stan, a package for Bayesian inference."""
+
+    homepage = "https://mc-stan.org/"
+    pypi     = "pystan/pystan-3.4.0.tar.gz"
+
+    maintainers = ['haralmha']
+
+    version('3.4.0', sha256='325e2fb0ab804555c05a603e0c9152ab11fcc3af01f3e9a9ff9fe9954b93184f')
+
+    depends_on('python@3.8:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('python@3.8:3', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    depends_on('py-setuptools', type='build')
+    depends_on('py-poetry-core@1.0.0:', type='build')
+    depends_on('py-aiohttp@3.6:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('py-aiohttp@3.6:3', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    depends_on('py-httpstan@:4.6,4.8:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('py-httpstan@4.7', type=('build', 'run'))
```
https://python-poetry.org/docs/dependency-specification/#tilde-requirements

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    depends_on('py-pysimdjson@3.2:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('py-pysimdjson@3.2:3', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    depends_on('py-numpy@1.19:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('py-numpy@1.19:1', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    depends_on('py-clikit@0.6:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('py-clikit@0.6:0', type=('build', 'run'))
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    homepage = "https://mc-stan.org/"
+    pypi     = "pystan/pystan-3.4.0.tar.gz"
+
+    maintainers = ['haralmha']
+
+    version('3.4.0', sha256='325e2fb0ab804555c05a603e0c9152ab11fcc3af01f3e9a9ff9fe9954b93184f')
+
+    depends_on('python@3.8:3', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('py-setuptools', type=('build', 'run'))
```
`pkg_resources` is needed at run-time.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-pystan/package.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    version('3.4.0', sha256='325e2fb0ab804555c05a603e0c9152ab11fcc3af01f3e9a9ff9fe9954b93184f')
+
+    depends_on('python@3.8:3', type=('build', 'run'))
+    depends_on('py-setuptools', type=('build', 'run'))
+    depends_on('py-poetry-core@1.0.0:', type='build')
+    depends_on('py-aiohttp@3.6:3', type=('build', 'run'))
+    depends_on('py-httpstan@4.7', type=('build', 'run'))
+    depends_on('py-pysimdjson@3.2:3', type=('build', 'run'))
+    depends_on('py-numpy@1.19:1', type=('build', 'run'))
+    depends_on('py-clikit@0.6:0', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pystan/package.py]:
```suggestion
    depends_on('py-clikit@0.6', type=('build', 'run'))
```
Caret requirements work differently if the major version is 0: https://python-poetry.org/docs/dependency-specification/#caret-requirements

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**